### ISSUE TITLE
ReplayGain support

### ISSUE THREAD

**Issue by jonathanlie** [kartik-venugopal/aural-player#12]
Hi, just a suggestion to improve list of things cleared up from README.

**Comment by maculateConception** [comment]
Hi, thanks for the suggestion. It has been suggested before, and I briefly looked into it. It is challenging to implement, but I will keep it in mind.

Cheers !

**Comment by kartik-venugopal** [comment]
@LesterJitsi - I have been looking into replay gain recently ... I can now implement replay gain reading from track metadata ... in the future, I may also be able to implement active scanning of tracks to compute replay gain (I found a couple of C / C++ libraries to do it). If I figure out how to integrate those libraries into Aural, I can do it.

Anyway, I should be able to release metadata-based replay gain in v4.0.0 😄

**Comment by LesterJitsi** [comment]
@kartik-venugopal  Great! 👍

**Comment by kartik-venugopal** [comment]
Implemented in the v4 "preview9" release (updated just now). Aural will look for metadata and if none is found, it will scan / analyze the file / album for loudness. There's also a pre-amp to adjust the baseline gain. It shows up as a new effects unit (on the effects window) ... enable it by clicking the power button on the unit.

![Screenshot 2024-08-25 at 2 00 30 AM](https://github.com/user-attachments/assets/510c935c-1034-46de-ae48-001c60ad3d3b)

See the release notes in the release Zip archive for all the details.

And let me know if it works for you. Cheers!

**Comment by LesterJitsi** [comment]
Yes, it works 😀 Now I'll reveal my ignorance posing a dumb question (please refer to the screenshots).

![Screenshot 2024-08-26 alle 18 25 35_1](https://github.com/user-attachments/assets/bd9b62d7-1c9c-4f58-ac6b-9a1da893e020)


![Screenshot 2024-08-26 alle 18 25 52_1](https://github.com/user-attachments/assets/fc6245b2-0b51-415d-b77a-3fbb07124000)

The tracks are taken from two Sheryl Crow's CDs mastered at markedly different loudness levels. No matter what the order of tracks in the Play Queue, the louder one is normalized to the less loud level (I hope my choice of words is not fully inaccurate). That's how it works ReplayGain or is it possible to reverse the effect? Disabling clipping prevention makes very little difference if none at all.

I'm asking just for sake of knowledge. I hate inflated loudness like this:

![stoogeswave2](https://github.com/user-attachments/assets/1c60f7a7-70bf-46f1-a840-1c0bccb2f862)

Thanks a lot!

**Comment by kartik-venugopal** [comment]
Hey @LesterJitsi , I think I understood what you're trying to say. Correct me if I didn't.

The way that ReplayGain generally works is based on the idea of a "target loudness" ... this is generally 89 dbSPL (Sound Pressure Level) or in the computer world, -18 LUFS (Loudness Units Full Scale) .. Aural uses -18 LUFS as its target loudness. Think of this value as a baseline ... anything louder than it will be made quieter, and anything softer than the baseline will be made louder. That is how track gain works. Album gain is a different story ... album gain is based on the average loudness for the entire album as a whole. But read the next paragraph ...

The way that Aural uses "Album gain" is ... first of all,  tracks must have the same value for the "Album" metadata field in order to be considered as part of the same album. If your  tracks in the Play Queue come from different albums (as in your case), then album gain will not be the right setting for you ...

That said, I can see from your screenshots that what has been applied is Track gain, not Album gain. So the target loudness is the deciding factor in what becomes louder / softer.

That said, maybe what Aural needs is to consider all the tracks currently in the Play Queue as an "album" ... maybe this would be another "Mode" ???

What do you think ?

**Comment by kartik-venugopal** [comment]
And clipping prevention won't take effect unless it predicts (based on track / album peak metadata) that clipping MAY occur ... if your loudness is already at a safe level, clipping will never happen and prevention makes no difference.

**Comment by LesterJitsi** [comment]
> That said, maybe what Aural needs is to consider all the tracks currently in the Play Queue as an "album" ... maybe this would be another "Mode" ???

Oh yes, that would be perfect 💯  You understood what I meant, I should have searched better. Ok also for clipping prevention, I mentioned it not being sure about what has to be considered a "safe level".

That said, you're doing an outstanding work 🥇 All paid alternatives to Aural for Mac I know about are less worthy, IMHO.

**Comment by kartik-venugopal** [comment]
Great, I am reopening this issue to add the new replay gain mode. Thanks buddy ... by the way, I hope you saw your name in the "Contributor Attributions" section on the main page 😉 

https://github.com/kartik-venugopal/aural-player?tab=readme-ov-file#contributor-attributions

**Comment by kartik-venugopal** [comment]
> > That said, maybe what Aural needs is to consider all the tracks currently in the Play Queue as an "album" ... maybe this would be another "Mode" ???
> 
> I mentioned it not being sure about what has to be considered a "safe level".

No worries. So, it's basically some mathematics ... it looks at the previously recorded "track peak" or "album peak" amplitude (which is a number between -1 and 1) ... and then based on the replay gain value (esp. if the gain is positive), it checks if the replay gain will cause the max amplitude to cross that threshold value beyond which the signal will get "clipped" or cut off, producing a distorted buzzy sound that we have all heard when blasting music on our stereo speakers 😄 

It's just a safety mechanism to prevent that unpleasant sound distortion that may occur ... whether or not it has an effect really depends on how much the amplitude varies from its quietest to loudest within a song / album.

This is a highly simplified explanation. Watch out for all the audiophiles who will come in here and correct me 😆
